File: 2021/10/31/gcd_of_strings.py
"""
https://leetcode.com/problems/greatest-common-divisor-of-strings/
Author: Andrew Jarombek
Date: 10/31/2021
"""


def gcd_of_strings(str1: str, str2: str) -> str:
    size = min(len(str1), len(str2))
    while size > 0:
        if len(str1) % size == 0 and len(str2) % size == 0 and str1[0:size] == str2[0:size]:
            str1_match = str1[0:size] * (len(str1) // size)
            str2_match = str2[0:size] * (len(str2) // size)

            if str1_match == str1 and str2_match == str2:
                return str1[0:size]
        size -= 1

    return ""

File: 2022/09/27/gcdOfStrings.ts
/**
 * https://leetcode.com/problems/greatest-common-divisor-of-strings/
 * @author Andrew Jarombek
 * @since 9/27/2022
 */

function gcdOfStrings(str1: string, str2: string): string {
    let result = str1.length > str2.length ? str2 : str1;

    while (result.length > 0) {
        if (
            str1.length % result.length == 0
            && str2.length % result.length == 0
            && str1 == str1.substring(result.length) + result
            && str2 == str2.substring(result.length) + result
        ) {
            return result;
        }

        result = result.slice(0, result.length - 1);
    }

    return result;
}